driver/mms.c

```diff
@@ -0,0 +1,430 @@
+    PetscCall(Problem1_GetData(pdata, t, xc[icell], yc[icell], DV_DY, &dvdy));
+    PetscCall(Problem1_GetData(pdata, t, xc[icell], yc[icell], DV_DT, &dvdt));
+
+    PetscCall(Problem1_GetData(pdata, t, xc[icell], yc[icell], N, &n));
+
+    PetscCall(Problem1_GetData(pdata, t, xc[icell], yc[icell], DZ_DX, &dzdx));
+    PetscCall(Problem1_GetData(pdata, t, xc[icell], yc[icell], DZ_DY, &dzdy));
+
+    PetscReal Cd = GRAVITY * (n * n) * PetscPowReal(h, -1.0 / 3.0);
+
+    h_source[icell] = dhdt + u * dhdx + h * dudx + v * dhdy + h * dvdy;
+
+    hu_source[icell] = u * dhdt + h * dudt;
+    hu_source[icell] += 2.0 * u * h * dudx + u * u * dhdx + GRAVITY * h * dhdx;
+    hu_source[icell] += u * h * dvdy + v * h * dudy + u * v * dhdy;
+    hu_source[icell] += dzdx * GRAVITY * h;
+    hu_source[icell] += Cd * u * PetscSqrtReal(u * u + v * v);
+
+    hv_source[icell] = v * dhdt + h * dvdt;
+    hv_source[icell] += u * h * dvdx + v * h * dudx + u * v * dhdx;
+    hv_source[icell] += v * v * dhdy + 2.0 * v * h * dvdy + GRAVITY * h * dhdy;
+    hv_source[icell] += dzdy * GRAVITY * h;
+    hv_source[icell] += Cd * v * PetscSqrtReal(u * u + v * v);
+
+  }
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+static PetscErrorCode Problem1_DirichletValue(ProblemData *pdata, PetscReal t, PetscInt nedges, PetscReal *xc_bnd_cell, PetscReal *yc_bnd_cell,
+                                              PetscReal *xc_edge, PetscReal *yc_edge, PetscReal *h_bnd, PetscReal *hu_bnd, PetscReal *hv_bnd,
+                                              PetscReal *bc_values) {
+  PetscFunctionBegin;
+
+  for (PetscInt iedge = 0; iedge < nedges; iedge++) {
+    PetscReal dx = xc_bnd_cell[iedge] - xc_edge[iedge];
+    PetscReal dy = yc_bnd_cell[iedge] - yc_edge[iedge];
+    PetscReal xc = xc_edge[iedge] - dx;
+    PetscReal yc = yc_edge[iedge] - dy;
+
+    PetscCall(Problem1_GetData(pdata, t, xc, yc, H, &h_bnd[iedge]));
+
+    PetscReal u_bnd, v_bnd;
+    PetscCall(Problem1_GetData(pdata, t, xc, yc, U, &u_bnd));
+    PetscCall(Problem1_GetData(pdata, t, xc, yc, V, &v_bnd));
+
+    hu_bnd[iedge] = h_bnd[iedge] * u_bnd;
+    hv_bnd[iedge] = h_bnd[iedge] * v_bnd;
+
+    bc_values[iedge * 3 + 0] = h_bnd[iedge];
+    bc_values[iedge * 3 + 1] = hu_bnd[iedge];
+    bc_values[iedge * 3 + 2] = hv_bnd[iedge];
+  }
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+int main(int argc, char *argv[]) {
+  // print usage info if no arguments given
+  if (argc < 2) {
+    usage(argv[0]);
+    exit(-1);
+  }
+
+  // initialize subsystems
+  PetscCall(RDyInit(argc, argv, help_str));
+
+  PetscMPIInt myrank, commsize;
+  PetscCallMPI(MPI_Comm_rank(PETSC_COMM_WORLD, &myrank));
+  PetscCallMPI(MPI_Comm_size(PETSC_COMM_WORLD, &commsize));
+
+  if (strcmp(argv[1], "-help")) {  // if given a config file
+
+    // create rdycore and set it up with the given file
+    MPI_Comm comm = PETSC_COMM_WORLD;
+    RDy      rdy;
+
+    PetscCall(RDyCreate(comm, argv[1], &rdy));
+
+    PetscCall(RDySetup(rdy));
```

Comment:
If we did make an alternate YAML format for MMS runs, we could add another version of this command (called `RDySetupMMS` or something) that would use a different (but similarly structured) input parser.